src/Development/IDE/Core/Rules.hs

```diff
@@ -334,11 +334,21 @@ instance Show GhcSessionFun where show _ = "GhcSessionFun"
 instance NFData GhcSessionFun where rnf !_ = ()
 
 
+-- Rule type for caching GHC sessions.
+type instance RuleResult GetHscEnvEq = HscEnvEq
+
+data GetHscEnvEq = GetHscEnvEq
```

Comment:
I'd have called this `GetHscEnv` since that's the real underlying type. The `Eq` is kind of a hack to get equality on it.

Reply:
Or why not just reuse the `ComponentOptions` type? Since all you ever do is convert it to that. If you want a newtype for the key (which isn't unreasonable) just `newtype GetHscEnv = GetHscEnv ComponentOptions` should be fine.

Reply:
Renamed to `GetHscEnv`.

Currently, only the `Main` module (and the `executable` component of this project) depends on hie-bios and I don't want to add that dependency to the `library` component just for this. And  because `ComponentOptions` comes from hie-bios, I'm not sure I can define instances of `Hashable`, `NFData`, and `Binary` for `GetHscEnv` in that case.

Reply:
Ah, good point about the dependencies. Your approach is definitely the right way to go.